components/layout/flow_ref.rs

```diff
@@ -8,194 +8,19 @@
-        }
-        WeakFlowRef { object: self.object }
-    }
-}
-
-impl<'a> Deref for FlowRef {
-    type Target = Flow + 'a;
-    fn deref(&self) -> &(Flow + 'a) {
-        unsafe {
-            mem::transmute_copy::<raw::TraitObject, &(Flow + 'a)>(&self.object)
-        }
-    }
-}
-
-impl DerefMut for FlowRef {
-    fn deref_mut<'a>(&mut self) -> &mut (Flow + 'a) {
-        unsafe {
-            mem::transmute_copy::<raw::TraitObject, &mut (Flow + 'a)>(&self.object)
-        }
-    }
-}
-
-impl Drop for FlowRef {
-    fn drop(&mut self) {
-        unsafe {
-            if self.object.vtable.is_null() ||
-               self.object.vtable as usize == mem::POST_DROP_USIZE {
-                return
-            }
-            if flow::base(&**self).strong_ref_count().fetch_sub(1, Ordering::Release) != 1 {
-                return
-            }
-            atomic::fence(Ordering::Acquire);
-
-            // Normally we'd call the underlying Drop logic but not free the
-            // allocation, but this is almost impossible without DST in
-            // Rust. Instead we make a fake trait object to run the drop logic
-            // on.
-            let flow_ref: FlowRef = mem::replace(self, FlowRef {
-                object: raw::TraitObject {
-                    vtable: ptr::null_mut(),
-                    data: ptr::null_mut(),
-                }
-            });
-
-            let vtable: &[usize; 3] = mem::transmute::<*mut (), &[usize; 3]>(flow_ref.object.vtable);
-            let object_size = vtable[1];
-            let object_align = vtable[2];
-
-            let fake_data = heap::allocate(object_size, object_align);
-            ptr::copy(flow_ref.object.data as *const u8, fake_data, object_size);
-
-            let fake_box = raw::TraitObject { vtable: flow_ref.object.vtable, data: fake_data as *mut () };
-            let fake_flow = mem::transmute::<raw::TraitObject, Box<Flow>>(fake_box);
-            drop(fake_flow);
-
-            if flow::base(&*flow_ref).weak_ref_count().fetch_sub(1, Ordering::Release) == 1 {
-                atomic::fence(Ordering::Acquire);
-                heap::deallocate(flow_ref.object.data as *mut u8, object_size, object_align);
-            }
-
-            mem::forget(flow_ref);
-        }
-    }
-}
-
-impl Clone for FlowRef {
-    fn clone(&self) -> FlowRef {
-        unsafe {
-            let _ = flow::base(&**self).strong_ref_count().fetch_add(1, Ordering::Relaxed);
-            FlowRef {
-                object: raw::TraitObject {
-                    vtable: self.object.vtable,
-                    data: self.object.data,
-                }
-            }
-        }
-    }
-}
-
-fn base<'a>(r: &WeakFlowRef) -> &'a BaseFlow {
-    let data = r.object.data;
-    debug_assert!(!data.is_null());
+// FIXME(https://github.com/servo/servo/issues/6503) This introduces unsound mutable aliasing.
+// Try to replace it with Arc::get_mut (which checks that the reference count is 1).
```

Comment:
I don't think you will succeed. The reference count will pretty much never be 1, since it's a doubly-linked tree.